cloud/resource_apikey.go

```diff
@@ -0,0 +1,353 @@
+	if expirationTime != "" {
+		if r1.MatchString(expirationTime) {
+			ago, err := str2duration.ParseDuration(expirationTime)
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("ERROR_PARSE_EXPIRATION_TIME: %w", err))
+			}
+			t = t.Add(ago)
+		} else if expirationTime != "No expiration date" {
+			layout := "2006-02-01T15:04:05Z"
+			t, err = time.Parse(layout, expirationTime)
+			if err != nil {
+				return diag.FromErr(fmt.Errorf("ERROR_PARSE_EXPIRATION_TIME: %w", err))
+			}
+		}
+	} else {
+		defaultExpireTime, err := time.ParseDuration("720h")
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("ERROR_PARSE_DEFAULT_EXPIRATION_TIME: %w", err))
+		}
+		t = t.Add(defaultExpireTime)
+	}
+	if expirationTime != "No expiration date" {
+		ak.Spec.ExpirationTime = &metav1.Time{Time: t}
+	}
+	if description != "" {
+		ak.Spec.Description = description
+	}
+	privateKey, err := util.GenerateEncryptionKey()
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_GENERATE_RSA_PRIVATE_KEY: %w", err))
+	}
+	encryptionKey, err := util.ExportPublicKey(privateKey)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_EXPORT_PUBLIC_KEY: %w", err))
+	}
+	ak.Spec.EncryptionKey = &v1alpha1.EncryptionKey{
+		PEM: encryptionKey.PEM,
+	}
+	revoke := d.Get("revoke").(bool)
+	ak.Spec.Revoke = revoke
+	_, err = clientSet.CloudV1alpha1().APIKeys(namespace).Create(ctx, ak, metav1.CreateOptions{
+		FieldManager: "terraform-create",
+	})
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_CREATE_API_KEY: %w", err))
+	}
+
+	d.SetId(fmt.Sprintf("%s/%s", namespace, name))
+	if err = d.Set(
+		"private_key", base64.StdEncoding.EncodeToString([]byte(util.ExportPrivateKey(privateKey)))); err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_SET_PRIVATE_KEY: %w", err))
+	}
+	err = retry.RetryContext(ctx, 10*time.Minute, func() *retry.RetryError {
+		dia := resourceApiKeyRead(ctx, d, m)
+		if dia.HasError() {
+			return retry.NonRetryableError(fmt.Errorf("ERROR_RETRY_CREATE_API_KEY: %s", dia[0].Summary))
+		}
+		ready := d.Get("ready")
+		if ready == "False" {
+			return retry.RetryableError(fmt.Errorf("CONTINUE_RETRY_CREATE_API_KEY"))
+		}
+		return nil
+	})
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_RETRY_CREATE_API_KEY: %w", err))
+	}
+	return nil
+}
+
+func resourceApiKeyDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	clientSet, err := getClientSet(getFactoryFromMeta(m))
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_INIT_CLIENT_ON_DELETE_API_KEY: %w", err))
+	}
+	namespace := d.Get("organization").(string)
+	name := d.Get("name").(string)
+	apikey, err := clientSet.CloudV1alpha1().APIKeys(namespace).Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("ERROR_READ_API_KEY: %w", err))
+	}
+	if apikey.Status.RevokedAt == nil {
+		return diag.FromErr(fmt.Errorf(
+			"ERROR_DELETE_API_KEY: no support delete apikey that not revoked, please revoke it first"))
```

Comment:
Why can the key that was not revoked not be deleted?

Reply:
Yes, this will prevent some objects from never being able to be cleaned up

Reply:
I think it's better to implement the webhook on API-server side to intercept the deletion behavior to throw error if exec delete action if the apikey is not revoked, wdyt? /cc @fantapsody @nlu90